src/framework/components/rigid-body/system.js

```diff
@@ -547,11 +567,19 @@ Object.assign(pc, function () {
                                 if (e0.collision) {
                                     e0.collision.fire("contact", forwardResult);
                                 }
+                                if (e0.rigidbody) {
+                                    e0.rigidbody.fire("contact", forwardResult);
+                                }
 
                                 // fire collisionstart events
                                 newCollision = this._storeCollision(e0, e1);
-                                if (newCollision && e0.collision) {
-                                    e0.collision.fire("collisionstart", forwardResult);
+                                if (newCollision) {
+                                    if (e0.collision) {
+                                        e0.collision.fire("collisionstart", forwardResult);
+                                    }
+                                    if (e0.rigidbody) {
```

Comment:
From inspection, it looks like your code will fire collisionstart from a rigidbody component even if the rigidbody component has no collisionstart handlers registered, right?

Reply:
That's correct Will, same as the collision element.

Reply:
Isn't that wrong? Why fire the event if there are no handlers registered?

Reply:
I guess, calling `hasEvent` (https://github.com/playcanvas/engine/blob/master/src/core/event-handler.js#L213-L215) has same price as calling `fire` (https://github.com/playcanvas/engine/blob/master/src/core/event-handler.js#L141-L143) if there is no event. As `fire` does same check as `hasEvent` method. But if there is a callback attached, then it is cheaper, as there will be one function call instead of two.

Reply:
Interesting. So we are probably better of removing those hasEvent checks and just calling fire for rigidbody+collision components that generates contacts/collisions.

Reply:
We still need to use _hasContactEvent because it's guarding more than event fire, it's also guarding the creation and storage of collision events.

Reply:
That's a very good point, yeah. OK, gonna approve.